[WIP]: Optimize Survey#to_csv by Eliminating N+1 Queries and Adding Foreign Key to SurveyNotification

Abishekcs opened the pull request:
TO BE UPDATED

## What this PR does
< describe the purpose of this pull request and note the issue it addresses >

## Screenshots
Before:
< add a screenshot of the UI before your change >

After:
< add a screenshot of the UI after your change >

ragesoss commented on app/models/surveys/survey.rb:
```diff
@@ -45,22 +45,29 @@ def status
 
+  def prepare_data_for_csv
+    respondents
+    course_for
```
This method names are not very clear about what they are for or what they do.

ragesoss commented on app/models/surveys/survey.rb:
```diff
@@ -86,13 +93,54 @@ def response(user)
+    notifications_by_user.each do |user_id, notification|
+      user = @users_by_id[user_id]
+      course = courses_by_id[notification.course_id] || fallback_course_for_user(user_id)
+      user.define_singleton_method(:survey_course) { course }
```
I think I understand what is going on here with `define_singleton_method`, but there must be a more elegant way to insert the course into the `notifications_by_user` data structure.

Some comments explaining the structure of this object (probably above `fetch_notifications_by_user`, since that's where the structure is created) would be helpful.

Abishekcs commented on app/models/surveys/survey.rb:
```diff
@@ -86,13 +93,54 @@ def response(user)
+  end
+
+  def load_survey_assignment_ids
+    @survey_assignment_ids = SurveyAssignment.where(survey_id: id).pluck(:id)
```
Okay, thank you! I was a bit confused about that—this really clears things up.

Abishekcs commented:
### Update Changes

- **Data Preparation Refactor:**
  -  Introduced `prepare_data_for_csv`, which initializes all required data before generating the CSV, avoiding repeated DB  queries.

  -  Replaced unclear method names like `respondents` and `course_for` with explicit, purpose-driven methods like 
     `survey_responding_users_by_id` and `course_slug_for_user.`

- **Improved Rails Conventions:**
   - Updated `survey_assignment_ids` to use conditional memoization following idiomatic Rails style for instance variable caching

- **More Efficient Course Mappings:**
  -  Replaced dynamic method definitions (`define_singleton_method`) with a cleaner approach using hash maps.
  -  Constructed a course slug lookup using `course_slug_for_user` and handled missing courses with a fallback (`fallback_course_slug_for_user`).

- **Commentary & Clarity:**
  -  Added inline comments explaining data structures like `notifications_by_user_id`, making it easier to understand object shapes and intentions.

### `prepare_data_for_csv` Call Flow

This method orchestrates the loading of all necessary data for efficient CSV generation:

1. **`survey_assignment_ids`**  
   Loads all survey assignment IDs for the current survey instance.

2. **`survey_responding_users_by_id`**  
   Fetches users who responded to the survey, indexed by their user ID.

3. **`notifications_by_user_id`**  
   Retrieves the primary `SurveyNotification` for each responding user.

4. **`courses_by_id`**  
   Loads course records relevant to the notifications.  
   This involves two sub-steps:
   
   4.1. **`notification_course_ids`**  
   Extracts unique course IDs from the notifications.
   
   4.2. **`load_courses_by_id(course_ids)`**  
   Fetches course details and indexes them by ID.


### **Peformance Metric:**


> **Note:** The screenshot below highlights the log for  
> `response_row = [user.username, course_slug_for_user(id)]`  
> while intentionally omitting `response(user)`, as that part is **not addressed in this PR**.  
> 
> The `response(user)` method still contains multiple N+1 queries, which remain a **major performance bottleneck** in the CSV export process which will be resolved in a separate PR.

**Before:**

![Before Code](https://github.com/user-attachments/assets/8e725007-698e-4e14-8702-44bb274f5b84)
![Before Screenshot](https://github.com/user-attachments/assets/c87f4066-07d4-4568-bd69-c1d9d769813f)


**After:**
![After Code ](https://github.com/user-attachments/assets/7abf3b45-2c5b-4348-bed3-ea198686298a)
![After Code](https://github.com/user-attachments/assets/e2abadea-fdf6-4c55-8bb9-c85e2ea4d2ab)

Abishekcs commented:
> For this, I'd like you to go through it with same strategy I encouraged for the other PR: see if you can find good-enough simpler optimizations that are based around just tweaking existing queries with `includes` before turning to the kinds of complex queries that return raw data structures. I'd really like to see each of these open PRs be as small as possible.

Okay

Abishekcs commented:
### Default one

![Deafult one (Main branch)](https://github.com/user-attachments/assets/1aea69ba-811e-4e05-9029-ba9d303daa66)


### implemented initially

![Initial Implementaion](https://github.com/user-attachments/assets/bb5ed20b-2904-410b-8b87-676c3c22236d)


### Latest Version

![Latest Version](https://github.com/user-attachments/assets/a7488d9e-2cc7-4e6a-8010-8bf6f8614e84)

ragesoss commented:
Great. The complex implementation was a little faster, and significantly reduced allocations, but this one gets most of the performance gain without adding to code complexity.
